### ISSUE TITLE
FR - Add Exclusion Tags for the No Hardlink checking

### ISSUE THREAD

**Issue by bakerboy448** [StuffAnThings/qbit_manage#18]
Title.  Add the ability to exclude certain tags that reside within the category to be checked for no hardlinks

Usecase: Permaseed MAM regardless of hardlink

Reason: *Arrs do not support per indexer post-download categories, and all my downloads get post-import dropped into `seeds`.  This is difficult to do and likely will not be coming ever.

Alt Solution:  Add support for per-indexer categories/management

**Comment by bobokun** [comment]
Can you please update to the latest [develop ](https://github.com/StuffAnThings/qbit_manage/tree/develop)branch and test?

**Comment by bakerboy448** [comment]
false positive?

cmd
`python qbit_manage.py -ro --tag-nohardlinks -ms --log-file /home/bakerboy448/qbit-mgr.log`

log https://gist.github.com/bakerboy448/0a6a4f0ded8876361df8da4f11594317

![image](https://user-images.githubusercontent.com/55419169/135207506-b3eb8aa4-f424-40dd-b2ac-ab7a02b39d4c.png)

```
bakerboy448@hetz-baker-2004:~$ ll /data/media/torrents/seeds/Diners.Drive-Ins.and.Dives.S07.720p.HULU.WEB-DL.DDP5.1.H.264-NTb
total 6626130
drwxrwxr-x   2 bakerboy448 media        15 Sep  6 17:21 ./
drwxrwxr-x 262 bakerboy448 media       419 Sep 28 17:41 ../
-rw-rw-r--   2 sonarr      media 537920093 Aug  3  2009 Diners.Drive-Ins.and.Dives.S07E01.Upper.Crust.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 538207380 Aug 10  2009 Diners.Drive-Ins.and.Dives.S07E02.Burgers.Tacos.and.Dogs.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537137700 Aug 17  2009 Diners.Drive-Ins.and.Dives.S07E03.Name.That.Cook.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537137153 Aug 24  2009 Diners.Drive-Ins.and.Dives.S07E04.Global.Grub.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 538021286 Aug 31  2009 Diners.Drive-Ins.and.Dives.S07E05.Only.at.This.Joint.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 538006650 Sep  7  2009 Diners.Drive-Ins.and.Dives.S07E06.You.Picked.Em.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 538178061 Oct  5  2009 Diners.Drive-Ins.and.Dives.S07E07.You.Can.Say.That.Again.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537250559 Oct 12  2009 Diners.Drive-Ins.and.Dives.S07E08.Barbecued.Baked.and.Brined.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537248562 Oct 19  2009 Diners.Drive-Ins.and.Dives.S07E09.Blast.from.the.Past.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537851782 Oct 26  2009 Diners.Drive-Ins.and.Dives.S07E10.Dives.Worth.a.Drive.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537795682 Nov  2  2009 Diners.Drive-Ins.and.Dives.S07E11.Family.Style.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537809060 Nov  9  2009 Diners.Drive-Ins.and.Dives.S07E12.Diners.Times.Three.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
-rw-rw-r--   2 sonarr      media 537921648 Nov 16  2009 Diners.Drive-Ins.and.Dives.S07E13.Legacies.720p.HULU.WEB-DL.DDP5.1.H.264-NTb.mkv
```

**Comment by bakerboy448** [comment]
filter seems to work tho  & a rerun fix it?

![image](https://user-images.githubusercontent.com/55419169/135208271-5cde93f8-99a2-4793-a306-afe4ac09a497.png)

cmd & log https://gist.github.com/bakerboy448/8530e14e73199d259cb30b111df6d800

**Comment by bobokun** [comment]
> false positive?
> 
> cmd `python qbit_manage.py -ro --tag-nohardlinks -ms --log-file /home/bakerboy448/qbit-mgr.log`
> 
> config https://gist.github.com/bakerboy448/a2776498b997fdb96e158ba42d26c771 log https://gist.github.com/bakerboy448/0a6a4f0ded8876361df8da4f11594317
> 
> ![image](https://user-images.githubusercontent.com/55419169/135207506-b3eb8aa4-f424-40dd-b2ac-ab7a02b39d4c.png)


That's weird, it shouldn't tag it as noHL. Can you sort by Name and see how many instances of the torrent are there? (If there are any cross-seeding). Can you also double check inside your qbittorrent webUI to see if the `save_path` in the general tab for that torrent is set to `/data/media/torrents/seeds/`.

I've also programmed it to automatically untag any noHL that were previously tagged if it finds any hard links. So can you try rerunning the command  `python qbit_manage.py --tag-nohardlinks --log-file /home/bakerboy448/qbit-mgr.log` to see if it removes the noHL tag. 

If that doesn't work can you try to manually untag the noHL from the torrent and rerun the command to see if it re-tags it falsely with noHL again?

**Comment by bakerboy448** [comment]
> Can you sort by Name and see how many instances of the torrent are there? (If there are any cross-seeding). 

None

> Can you also double check inside your qbittorrent webUI to see if the save_path in the general tab for that torrent is set to /data/media/torrents/seeds/.

confirmed 
`Save Path: | /data/media/torrents/seeds/`

> Also if you aren't using docker you can try to remove the remote_dir variable completely in your config to see if that helps fix the issue.

Correct, am native.
Commented out ` #remote_dir: ''`

No change.

Removed the tag manually and reran (including the change); it only did 49 torrents - a much more reasonable number.
Sample of 1 confirmed only 1 link - so appears to be working as expected.


> I've also programmed it to automatically untag any noHL that were previously tagged if it finds any hard links. So can you try rerunning the command python qbit_manage.py --tag-nohardlinks --log-file /home/bakerboy448/qbit-mgr.log to see if it removes the noHL tag.
> If that doesn't work can you try to manually untag the noHL from the torrent and rerun the command to see if it re-tags it falsely with noHL again?

This is what was done prior to opening the issue and what created the logs :) They did falsely get retagged.


So TL;DR
- `remote_dir` breaks functionality on non-docker
- removal of tagged torrents does not appear to function (based on uncommenting remote_dir and it not removing any from the tag) (manual tag removal and rerun it appears ~400 torrents should have been removed)

**Comment by bobokun** [comment]
> So TL;DR
> 
> * `remote_dir` breaks functionality on non-docker

Thanks for the detailed report. I'll make a check for `remote_dir: ''` in case it's being defined as empty it shouldn't break this function. 

> * removal of tagged torrents does not appear to function (based on uncommenting remote_dir and it not removing any from the tag) (manual tag removal and rerun it appears ~400 torrents should have been removed)

This seems to be a bug. It should've untagged noHL without you having to manually remove the tag.

**Comment by bobokun** [comment]
Can you please update again to the latest [develop ](https://github.com/StuffAnThings/qbit_manage/tree/develop)branch again?
I've fixed it so now you can use your original config where `remote_dir: ''` shouldn't break it. 

Also you can test by manually tag torrents **with** hard links as noHL in qbit and run the script again, it should now remove the noHL tag this time. Please let me know if you find any additional bugs, if not I'll migrate the changes to master branch. Thanks!

**Comment by bakerboy448** [comment]
ran with remote_dir commented out

it ignored the ignore tag for mam
```

2021-09-29 12:38:21,559 - INFO:
 - Torrent Name: Demigods and Magicians has no hard links found. Adding tags noH                                                                   L.
 - Torrent Name: Steven Rinella - Meat Eater; Adventures in the Life of an Ameri                                                                   can Hunter has no hard links found. Adding tags noHL.
 - Torrent Name: The Complete Guide to Hunting, Butchering, and Cooking Wild Gam                                                                   e, Volume 2 = Small Game and Fowl @ Steven Rinella has no hard links found. Addi                                                                   ng tags noHL.
 - Torrent Name: Catching Fire has no hard links found. Adding tags noHL.
 - Torrent Name: Sea of Monsters by Ranson Riordan has no hard links found. Addi                                                                   ng tags noHL.
 - Torrent Name: The Complete Guide to Hunting, Butchering, and Cooking Wild Gam                                                                   e, Volume 1 = Big Game @ Steven Rinella has no hard links found. Adding tags noH                                                                   L.
 - Torrent Name: The Washington Post - The Original Watergate Stories has no har                                                                   d links found. Adding tags noHL.
 - Torrent Name: Harry Potter Books 1-7 has no hard links found. Adding tags noH                                                                   L.
 - Torrent Name: Percy Jackson and the Olympians #1-5 has no hard links found. A                                                                   dding tags noHL.
 - Torrent Name: For Magnus Chase - Hotel Valhalla, Guide to the Norse Worlds ha                                                                   s no hard links found. Adding tags noHL.
 - Torrent Name: The_MeatEater_Fish_and_Game_Cookbook_-_Steven_Rinella has no ha                                                                   rd links found. Adding tags noHL.
2021-09-29 12:38:21,560 - INFO: tag/set ratio limit/seeding time for  11 .torren                                                                   ts(s)
```

-----

- Added a hardlinked torrent to nlHL
- uncommented remote_dir
- re-ran
-  added all torrents; did not ignore remote_dir & did not untag the torrent with links
`2021-09-29 12:40:13,856 - INFO: tag/set ratio limit/seeding time for  460 .torrents(s)`
- recommented remote_dir
- re-ran
- did not untag
